qiskit_qec/operators/base_xp_pauli.py

```diff
@@ -397,6 +518,178 @@ def _append_circuit(
+
+    def unique_vector_rep(self):
+        """_summary_"""
+        return self._unique_vector_rep()
+
+    def _unique_vector_rep(self):
+        """(TODO improve doc): This is the equivalent of XPRound from Mark's
+        code. It converts the XPPauli operator into unique vector form, ie
+        phase_exp in Z modulo 2*precision, x in Z_2, z in Z modulo
+        precision."""
+        matrix = np.empty(shape=np.shape(self.matrix), dtype=np.int64)
+
+        phase_exp = np.mod(self._phase_exp, 2 * self.precision)
+        matrix[:, : self.num_qubits] = np.mod(self.x, 2)
+        matrix[:, self.num_qubits :] = np.mod(self.z, np.expand_dims(self.precision, axis=-1))
+
+        return BaseXPPauli(matrix, phase_exp, self.precision)
+
+    def rescale_precision(self, new_precision):
+        """_summary_"""
+        return self._rescale_precision(new_precision)
+
+    def _rescale_precision(self, new_precision):
+        """(TODO improve doc): This is the equivalent of XPSetNsingle from
+        Mark's code. It rescales the generalized symplectic vector components
+        of XPPauli operator to the new precision. Returns None if the
+        rescaling is not possible, else returns the rescaled BaseXPPauli object."""
+
+        # TODO Currently, if any operator in an XPPauliList can not be
+        # rescaled, this function will return None.
+        unique_xp_op = self.unique_vector_rep()
+        old_precision = np.atleast_1d(unique_xp_op.precision)
+        new_precision = np.atleast_1d(new_precision)
+        cmp = new_precision > old_precision
+        matrix = np.empty(shape=np.shape(unique_xp_op.matrix), dtype=np.int64)
+        phase_exp = np.empty(shape=np.shape(unique_xp_op._phase_exp))
+
+        for i, val in enumerate(cmp):
+
+            if val:
+                if np.mod(new_precision[i], old_precision[i] > 0):
+                    return None
+                scale_factor = new_precision[i] // old_precision[i]
+                phase_exp[i] = scale_factor * unique_xp_op._phase_exp[i]
+                matrix[:, unique_xp_op.num_qubits :][i] = (
+                    scale_factor * np.atleast_2d(unique_xp_op.z)[i]
+                )
+
+            else:
+                scale_factor = old_precision[i] // new_precision[i]
+                if (
+                    (old_precision[i] % new_precision[i] > 0)
+                    or (np.sum(np.mod(unique_xp_op._phase_exp[i], scale_factor)) > 0)
+                    or (np.sum(np.mod(unique_xp_op.z[i], scale_factor)) > 0)
+                ):
+                    return None
+                phase_exp[i] = unique_xp_op._phase_exp[i] // scale_factor
+                matrix[:, unique_xp_op.num_qubits :][i] = (
+                    np.atleast_2d(unique_xp_op.z)[i] // scale_factor
+                )
+
+        matrix[:, 0 : unique_xp_op.num_qubits] = unique_xp_op.x
+
+        return BaseXPPauli(matrix, phase_exp, new_precision)
+
+    def weight(self):
+        """_summary_"""
```

Comment:
Add docs and checks etc

Reply:
@dsvandet is there any particular check which you think should be added here?